### PR TITLE
Callum changes -- working sidebar & local data storage state.

### LINKED ISSUE #4: User word list sidebar content locked to specific tabs
title
opin?

### PR DESCRIPTION
**Pull request by CallumBeaney**
- user word list and page data now held in local storage
- sidebar now built from local storage
- sidebar building done based on page (URL)
- IDs for modal buttons/tab-buttons made longer to prevent security issues
- cleaned up all functionality, added comments etc

tryCatching etc needs doing.
So does adding e.g. remove buttons for sidebar, various option bells & whistles.

Important. Some of this, particularly state-checking if()s in  `sidebar.js` is more procedural than I would like. It's like that for a reason. JSbad etc.

### CODE REVIEW

**Comment by CallumBeaney** [comment]
Done try catch, and also added some utility fns for error logging & handling.
There's no good reason they will ever actually be triggered.

As much of this is "if it breaks, everything breaks", they log the error with the context passed into them, and return from the functions in question. 

Perhaps we can create a custom modal with a message asking to log the error by copying their console into a GH issue.

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 todo.txt
 peopleToReachOutTo.txt
 /OCR/*
+lem.js
```
Comment (alexobviously) [inline .gitignore]: wat is

File: src/background.js [hunk 2]
```diff
@@ -1,16 +1,37 @@
...
+  
+  Background Scripts: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts
+  Local Storage:      https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/local
+  Context Menus:      https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus.
 */
 
-/// STORAGE.LOCAL: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/local
-/// CONTEXT MENUS: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus
-
 ! async function setStateFirstTime(){
```
Comment (alexobviously) [inline src/background.js]: `writeInitialState` maybe?

File: src/globals.js [hunk 3]
```diff
@@ -112,71 +143,73 @@ class MatchedWordEntry {
...
 
 // FUNNY FONTS, MARGINALIA IMAGE HANDLING &c.
-// this is an ugly way of doing this but just for proof of concept this is *one* way.
 
 /**
  * @summary this is altered in content.js -> dictionaryEntriesToHTMLtext() when creating the sidebar. It is instantiated when the user first opens a sidebar, after which it will not be changed
  */
 var persistentSideBarMarginaliaURL = null;
 
 const marginaliaFilepaths = [
```
Comment (alexobviously) [inline src/globals.js]: `marginaliaIds`?

File: src/globals.js [hunk 4]
```diff
@@ -112,71 +143,73 @@ class MatchedWordEntry {
...
+  "threeFish",
+  "toad",
+  "unicorn",
+  "unicorn2",
+  "wolf",
+  "winter",
+  "wolf"
 ];
 
 function getRandomImagePath() {
```
Comment (alexobviously) [inline src/globals.js:243]: maybe jsdoc this?

File: src/helpers.js [hunk 5]
```diff
@@ -5,31 +5,34 @@
...
+  }
 }
 
-
-function addPageToUserPagesList() {
-  const url = extractBaseURLOfPage();
-  const urlExists = userPages.hasOwnProperty(url);
-  if (!urlExists) { 
-    userPages[url] = buildPageInfo();
+function stateError(context, state) {
```
Comment (alexobviously) [inline src/helpers.js:49]: jsdoc this for return type etc

File: src/helpers.js [hunk 6]
```diff
@@ -5,31 +5,34 @@
...
- */
-function addWordToUserList(word) {
-  const hasCommonIndex = userAddedWords.some((e) => e.lookupIndex === word.lookupIndex && e.url === word.url);
-  if (hasCommonIndex) return; 
-  
-  userAddedWords.push(word);
-  addPageToUserPagesList();
-  updateSidebar();
-  // console.log("addWordToUserList : added word: " + word.matchedVariant + " id: " + word.lookupIndex);
+// "If managed storage is not set, undefined will be returned."
```
Comment (alexobviously) [inline src/helpers.js]: might as well document this with jsdoc

File: src/helpers.js [hunk 7]
```diff
@@ -64,6 +66,90 @@ function htmlize(entry) {
...
+}
+
+
+
+/*
+    ,-----------.
+   (_\           \
+      |           |
+      |           |
+      |  js bad   |
```
Comment (alexobviously) [inline src/helpers.js:160]: agre

**Review by alexobviously — APPROVED** [review]
looks good

**Pull request merged**